engine/resources/pippet.go

```diff
@@ -0,0 +1,183 @@
+}
+
+// PippetRes is a wrapper resource for puppet.
+// It implements the functional equivalent of an exec resource
+// that calls "pupept resource <type> <title> <params>", but offers
+// superior performance through a long-running Puppet process that
+// receives resources through a pipe (hence the name).
+type PippetRes struct {
+	traits.Base // add the base methods without re-implementation
+	traits.Refreshable
+
+	init *engine.Init
+
+	Type   string `yaml:"type" json:"type"`
+	Title  string `yaml:"title" json:"title"`
+	Params string `yaml:"params" json:"params"`
+
+	runner PippetRunner
+}
+
+// Default returns an example Pippet resource.
+func (obj *PippetRes) Default() engine.Res {
+	return &PippetRes{
+		Params: "{}",
+	}
+}
+
+// Validate never errors out. We don't know the set of potential
+// types that Puppet supports. Resource names are arbitrary. We
+// cannot really validate the parameter YAML, because we cannot
+// assume that it can be unmarshalled into a map[string]string.
+func (obj *PippetRes) Validate() error {
+	return nil
+}
+
+// Init makes sure that the PippetRunner object is initialized.
+func (obj *PippetRes) Init(init *engine.Init) error {
+	obj.init = init // save for later
+
+	var err error
+	obj.runner, err = PippetRunnerInstance()
+	if err != nil {
+		return err
+	}
+
+	obj.runner.Register(obj)
+
+	return nil
+}
+
+// Close is run by the engine to clean up after the resource is done.
+func (obj *PippetRes) Close() error {
+	obj.runner.Unregister(obj)
+	return nil
+}
+
+// Watch is the primary listener for this resource and it outputs events.
+func (obj *PippetRes) Watch() error {
+	obj.init.Running() // when started, notify engine that we're running
+
+	select {
+	case <-obj.init.Done: // closed by the engine to signal shutdown
+	}
+
+	//obj.init.Event() // notify engine of an event (this can block)
+
+	return nil
+}
+
+// CheckApply synchronizes the resource if required.
+func (obj *PippetRes) CheckApply(apply bool) (checkOK bool, err error) {
+	if obj.init.Refresh() {
+		obj.init.Logf("WARNING - pippet does not support refreshing yet")
+	}
+	result := obj.runner.Apply(obj.Type, obj.Title, obj.Params)
+	if result.Error {
+		return false, fmt.Errorf("pippet: %s[%s]: ERROR - %s", obj.Type, obj.Title, result.Exception)
+	}
+	if result.Failed {
+		return false, fmt.Errorf("pippet: %s[%s]: apply failed - %s", obj.Type, obj.Title, result.Exception)
+	}
+	return !result.Changed, nil // state is always okay
+}
+
+// Cmp compares two resources and returns an error if they are not equivalent.
+func (obj *PippetRes) Cmp(r engine.Res) error {
+	res, ok := r.(*PippetRes)
+	if !ok {
+		return fmt.Errorf("not a %s", obj.Kind())
+	}
+
+	if obj.Type != res.Type {
+		return fmt.Errorf("the resource types differ")
+	}
+
+	if obj.Title != res.Title {
+		return fmt.Errorf("the resource titles differ")
+	}
+
+	// FIXME: This is a lie. Parameter lists can be equivalent
```

Comment:
You're allowed to do a more complex Cmp if it's appropriate. Is this what you mean? Eg: compare the sorted lists that you mapped with Trim?

Reply:
I don't think it's that simple. Given two arbitrary YAML hashes, you will be hard pressed to make a call on their being equivalent without actually parsing the YAML. And Go cannot casually do this.